broker/query_plan.go

```diff
@@ -35,75 +46,138 @@ func (bpn *blockingPlanNodeImpl) Add(nodes ...common.BlockingPlanNode) {
 	bpn.children = append(bpn.children, nodes...)
 }
 
-// MergeNode is a blocking plan node that merges results from sub nodes
-type MergeNode struct {
+func NewMergeNode(agg common.AggType) common.MergeNode {
+	return &mergeNodeImpl{
+		aggType: agg,
+	}
+}
+
+type mergeNodeImpl struct {
 	blockingPlanNodeImpl
 	// MeasureType decides merge behaviour
-	AggType common.AggType
+	aggType common.AggType
 }
 
-func (mn *MergeNode) Run() (result queryCom.AQLQueryResult, err error) {
+func (mn *mergeNodeImpl) AggType() common.AggType {
+	return mn.aggType
+}
+
+func (mn *mergeNodeImpl) Run(ctx context.Context) (result queryCom.AQLQueryResult, err error) {
 	nChildren := len(mn.children)
 	// checks before fan out
-	if common.Avg == mn.AggType {
+	if common.Avg == mn.aggType {
 		if nChildren != 2 {
 			err = utils.StackError(nil, "Avg MergeNode should have 2 children")
 			return
 		}
-		lhs, ok := mn.children[0].(*MergeNode)
+		lhs, ok := mn.children[0].(common.MergeNode)
 		if !ok {
 			err = utils.StackError(nil, "LHS of avg node must be sum node")
 			return
 		}
-		if common.Sum != lhs.AggType {
+		if common.Sum != lhs.AggType() {
 			err = utils.StackError(nil, "LHS of avg node must be sum node")
 			return
 		}
-		rhs, ok := mn.children[0].(*MergeNode)
+		rhs, ok := mn.children[1].(common.MergeNode)
 		if !ok {
 			err = utils.StackError(nil, "RHS of avg node must be count node")
 			return
 		}
-		if common.Count != rhs.AggType {
+		if common.Count != rhs.AggType() {
 			err = utils.StackError(nil, "RHS of avg node must be count node")
 			return
 		}
 	}
 
 	childrenResult := make([]queryCom.AQLQueryResult, nChildren)
-	errs := make([]error, nChildren)
+	errs := 0
 	wg := &sync.WaitGroup{}
 	for i, c := range mn.children {
 		wg.Add(1)
-		bpn, ok := c.(common.BlockingPlanNode)
-		if !ok {
-			err = utils.StackError(nil, "expecting BlockingPlanNode")
-			return
-		}
 		go func(i int, n common.BlockingPlanNode) {
 			defer wg.Done()
 			var res queryCom.AQLQueryResult
-			res, err = n.Run()
+			res, err = n.Run(ctx)
 			if err != nil {
 				// err means downstream retry failed
-				errs[i] = utils.StackError(err, "child node failed")
+				utils.GetLogger().With(
+					"error", err,
+				).Error("child node failed")
+				errs++
 				return
 			}
 			childrenResult[i] = res
-		}(i, bpn)
+		}(i, c)
 	}
 	wg.Wait()
 
+	if errs > 0 {
+		err = utils.StackError(nil, fmt.Sprintf("%d errors happend running merge node", errs))
+		return
+	}
+
 	result = childrenResult[0]
 	for i := 1; i < nChildren; i++ {
-		ctx := newResultMergeContext(mn.AggType)
-		result = ctx.run(result, childrenResult[i])
-		if ctx.err != nil {
-			err = ctx.err
+		mergeCtx := newResultMergeContext(mn.aggType)
```

Comment:
future improve: early merge before waiting for all result come back

Reply:
yeah will add todo